erc20.go

```diff
@@ -0,0 +1,363 @@
+	"github.com/threefoldtech/rivine/modules"
+	"github.com/threefoldtech/rivine/persist"
+	"github.com/threefoldtech/rivine/pkg/encoding/rivbin"
+	rivinetypes "github.com/threefoldtech/rivine/types"
+
+	bolt "github.com/rivine/bbolt"
+)
+
+const (
+	pluginDBVersion = "1.0.0.0"
+	pluginDBHeader  = "erc20Plugin"
+)
+
+var (
+	// buckets for the ERC20-bridge feature
+	bucketERC20ToTFTAddresses = []byte("addresses_erc20_to_tft") // erc20 => TFT
+	bucketTFTToERC20Addresses = []byte("addresses_tft_to_erc20") // TFT => erc20
+	bucketERC20TransactionIDs = []byte("erc20_transactionids")   // stores all unique ERC20 transaction ids used for erc20=>TFT exchanges
+)
+
+type (
+	// Plugin is a struct defines the minting plugin
+	Plugin struct {
+		storage            modules.PluginViewStorage
+		unregisterCallback modules.PluginUnregisterCallback
+	}
+
+	transactionContext struct {
+		BlockHeight  rivinetypes.BlockHeight
+		BlockTime    rivinetypes.Timestamp
+		TxSequenceID uint16
+	}
+)
+
+// NewERC20Plugin creates a new erc20 plugin
+func NewERC20Plugin(ERC20FeePoolAddress rivinetypes.UnlockHash, oneCoin rivinetypes.Currency, erc20TxValidator types.ERC20TransactionValidator) *Plugin {
+	p := &Plugin{}
+	rivinetypes.RegisterTransactionVersion(types.TransactionVersionERC20Conversion, types.ERC20ConvertTransactionController{})
+	rivinetypes.RegisterTransactionVersion(types.TransactionVersionERC20CoinCreation, types.ERC20CoinCreationTransactionController{
+		Registry:    p,
+		OneCoin:     oneCoin,
+		TxValidator: erc20TxValidator,
+	})
+	rivinetypes.RegisterTransactionVersion(types.TransactionVersionERC20AddressRegistration, types.ERC20AddressRegistrationTransactionController{
+		Registry:             p,
+		OneCoin:              oneCoin,
+		BridgeFeePoolAddress: ERC20FeePoolAddress,
+	})
+	return p
+}
+
+// InitPlugin initializes the Bucket for the first time
+func (p *Plugin) InitPlugin(metadata *persist.Metadata, bucket *bolt.Bucket, storage modules.PluginViewStorage, unregisterCallback modules.PluginUnregisterCallback) (persist.Metadata, error) {
+	p.storage = storage
+	p.unregisterCallback = unregisterCallback
+	if metadata == nil {
+		buckets := [][]byte{
+			bucketERC20ToTFTAddresses,
+			bucketTFTToERC20Addresses,
+			bucketERC20TransactionIDs,
+		}
+		var err error
+		for _, bckt := range buckets {
+			_, err = bucket.CreateBucketIfNotExists(bckt)
+			if err != nil {
+				return persist.Metadata{}, err
+			}
+		}
+
+		metadata = &persist.Metadata{
+			Version: pluginDBVersion,
+			Header:  pluginDBHeader,
+		}
+	} else if metadata.Version != pluginDBVersion {
+		return persist.Metadata{}, errors.New("There is only 1 version of this plugin, version mismatch")
+	}
+	return *metadata, nil
+}
+
+// ApplyBlock applies a block's erc20 transaction to their buckets.
+func (p *Plugin) ApplyBlock(block rivinetypes.Block, height rivinetypes.BlockHeight, bucket *persist.LazyBoltBucket) error {
+	if bucket == nil {
```

Comment:
should not be required though? Why would this function ever be called with a `nil` bucket? can that happen from the rivine side? You could also turn it into an interface of course so that the writer of a plugin doesn't get scared of the pointer.